### PR TITLE
RavenDB-17805 - IndexFieldMapping API to bind analyzers with indexing fields

### PR DESCRIPTION
**Pull request by redknightlois**
The current implementation was started as part of the proof of concept from @ayende into the MoreLikeThis feature. The Index Field Mapping API requires its own PR with its own review on it. 

https://issues.hibernatingrhinos.com/issue/RavenDB-17805

### CODE REVIEW

File: src/Corax/IndexWriter.cs [hunk 1]
```diff
@@ -29,14 +30,106 @@ public enum TermIdMask : long
...
+
+        public IndexFieldsMapping(ByteStringContext context)        
+        {
+            _context = context;
+            _fields = new Dictionary<Slice, IndexFieldBinding>();
+            _fieldsById = new Dictionary<int, IndexFieldBinding>();
+            _fieldsList = new List<IndexFieldBinding>();
+        }
+
+        public IndexFieldsMapping AddBinding(int fieldId, Slice fieldName, Analyzer analyzer = null)
```
Comment (ayende) [inline src/Corax/IndexWriter.cs:53]: Why do we need to provide `fieldId` externally? Shouldn't we just use `_fields.Count+1` ?

**Review by ayende — COMMENTED** [review]
In general, I can't review Corax PRs at this point.
There is too much overlap between PRs, too much intermingling between things and I get lost.

@arekpalinski @redknightlois @maciejaszyk 

We *have* to merge all the outstanding PRs and continue from a known state.

File: src/Corax/IndexWriter.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Corax/IndexWriter.cs:53]: I've modified `IndexDefiniton` to have also `fieldId`. This way we've confidence that ids are always identical.

Comment (redknightlois) [reply src/Corax/IndexWriter.cs:53]: It is important for Corax to have `fieldId` if RavenDB then decides to use sequential and do not store it is fine. But the internal of Corax uses those `fieldId` values explicitely.

**Comment by redknightlois** [comment]
@arekpalinski Tests pass now, but I think there are a few improvements that should be done to the integration. I will setup a call with @maciejaszyk before I am out on vacations next week to talk about the improvements we can make.

**Comment by arekpalinski** [comment]
There are still few Corax test failures. @maciejaszyk could you take a look?

**Comment by arekpalinski** [comment]
Fixed in https://github.com/ravendb/ravendb/pull/13430

**Comment by arekpalinski** [comment]
Closing this PR

**Pull request closed without merging**